### ISSUE TITLE
Adding Post Message Support on WebGL for Cross-Origin Comms

### ISSUE THREAD

**Issue by MoizAhmedd** [gree/unity-webview#1088]
Currently, the only way to communicate with the webview on WebGL, is using EvaluateJS

This is problematic, because running`.eval` on the iframe that the webview creates, is blocked by browser when applied on cross-origin frames.

A safe and common way to get around this is to post messages to the iframe instead.

It would be great for the WebGL version to have support for post message, so that communication between game and cross-origin frames can occur

**Comment by KojiNakamaru** [comment]
I pushed a modified sample app to introduce postMessage:

https://github.com/gree/unity-webview/commits/example-for-1088/

As we can see in https://github.com/gree/unity-webview/commit/40aba92f1b6788c852732bf176af76c539fd7aae, postMessage requires addEventListener in the iframe content, so it is not a complete alternative. Anyway, as shown here, adding postMessage is easy and might be adequate if you can edit the iframe content.

**Comment by MoizAhmedd** [comment]
looks pretty good, will test this out

**Comment by KojiNakamaru** [comment]
Yes, so I wrote "not a complete alternative."

**Comment by jonobarel-slatescience** [comment]
@KojiNakamaru do I understand correctly the change in https://github.com/gree/unity-webview/commit/d2c9e5ba76fbbfce19f4ad417aa87f9520b07bf9 - does this mean if I branch off this revision, I can directly utilize postmessage via existing calls to EvaluateJS() (line 1120 below)?

https://github.com/gree/unity-webview/blob/d2c9e5ba76fbbfce19f4ad417aa87f9520b07bf9/sample/Assets/Plugins/WebViewObject.cs#L1115-L1122

**Comment by KojiNakamaru** [comment]
@jonobarel-slatescience 
Yes. sample.html is modified as below:

https://github.com/gree/unity-webview/compare/2ce0ca1...d2c9e5b#diff-4497f115d85aa6ac45420130898f9b46e85b1afe7683a050f27551961e460f52

so it receives and evaluates a message from EvaluateJS.

**Comment by jonobarel-slatescience** [comment]
Okay, I see m mistake - you've created a dedicated version of the plugin and WebViewObject.cs inside the `sample` folder.

I'm copying the changes from there into the main plugin to test a build of our app. Thanks!

**Comment by jonobarel-slatescience** [comment]
@KojiNakamaru FYI I was able to test this and confirm that it works, but I will be refactoring a bit for clarity and post a PR if you're okay with that.

**Comment by KojiNakamaru** [comment]
@jonobarel-slatescience 
Thank you, I'll appreciate it as a nice reference for others. The PR however might be difficult to be merged to the master as it requires the html content to define an event listener.

**Comment by jonobarel-slatescience** [comment]
Fair point. I've got a new build and we'll be testing with the changes.

We will likely integrate the changes into our fork, if they continue to work.

Note that I have this block I'm unsure about, so for WebGL I've excluded it.

https://github.com/SlateScience/unity-webview/blob/1f34e625dcce8454c954c2f9a5dbc50c5c9cbc3c/dist/package/Assets/Plugins/WebViewObject.cs#L944-L949

**Comment by KojiNakamaru** [comment]
> ```
> #if !UNITY_WEBGL
>         if (GetVisibility() && !v)
>         {
>             EvaluateJS("if (document && document.activeElement) document.activeElement.blur();");
>         }
> #endif
> ```
> https://github.com/SlateScience/unity-webview/blob/1f34e625dcce8454c954c2f9a5dbc50c5c9cbc3c/dist/package/Assets/Plugins/WebViewObject.cs#L944-L949

I agree that this modification is reasonable, as completely avoiding EvaluateJS() allows to display html sites in different domains without SecurityError. At least this modification might be better to be integrated to the master.

**Comment by jonobarel-slatescience** [comment]
Hi @KojiNakamaru 

Thanks again for your ongoing support on this.

In my implementation, I use a fallback for non-WebGL that reverts to using EvaluateJS.

**Comment by KojiNakamaru** [comment]
I'm a bit confused. Originally postMessage is required only for UNITY\_WEBGL to implement EvaluateJS(). Do you need to define public PostMessage() for other tasks? If this understanding is okay, how about the following:

```c#
    public void PostMessage(string msg)
    {
#if UNITY_WEBGL
        ...
#else
        EvaluteJS($"window.postMessage('{msg}', '*')");
#endif
    }
```

**Comment by jonobarel-slatescience** [comment]
That's exactly how I implemented it.

Our concern was that a native webview (ios/android), when loading a page with CSP (Content Security Policy) that disallows `eval` calls, would cause errors.

We found that on iOS and Android, there are no errors when using eval on pages with a CSP header _that excludes_ `unsafe-eval`, the calls still work. 

Could it be that these webviews don't enforce CSP? Only when loaded through a browser? do you know if that's the case?

**Comment by KojiNakamaru** [comment]
The implementations for Android/iOS utilize native WebView/WKWebView APIs to evaluate javascript strings, so that they are not affected by CSP.

**Comment by jonobarel-slatescience** [comment]
Thanks, so if I understand correctly, any `eval` call that directly goes through the native API, will ignore the CSP.
But eval calls _in the page_ will still be enforced?

**Comment by KojiNakamaru** [comment]
I understand so, though I've not actually experimented it in detail.

**Comment by jonobarel-slatescience** [comment]
I can't find it in the documentation, but [this discussion](https://github.com/w3c/webappsec-csp/issues/570) concurs with you.

I think we're good with it this way. Thanks again! (not my issue, but I consider it closed).
